### PR TITLE
Use for-of loops instead of forEach

### PR DESCRIPTION
**Pull request by ooflorent**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

no

**If relevant, link to documentation update:**

n/a

**Summary**

Use `for-of` loops instead of `forEach` loops to reduce memory usage in intensive builds.
Here below are some data recorded before and after this PR.

```
node v8.9.3
  before:
    build times (ms): 872, 854, 841, 845, 845, 874, 858, 863, 857, 855
    mean (ms): 856
    median (ms): 856
    used memory (MB): 33.8, 41.3, 39.5, 38.9, 36.9, 30.1, 32.8, 39.6, 31.3, 32.4
    mean (MB): 35.7
    median (MB): 35.3
  after:
    build times (ms): 882, 882, 897, 884, 887, 865, 856, 858, 867, 878
    mean (ms): 876
    median (ms): 880
    used memory (MB): 33.6, 33.7, 34.0, 33.8, 33.7, 27.2, 33.9, 33.8, 33.6, 33.6
    mean (MB): 33.1
    median (MB): 33.7

node v9.4.0
  before:
    build times (ms): 811, 804, 827, 814, 895, 823, 849, 853, 853, 835
    mean (ms): 836
    median (ms): 831
    used memory (MB): 39.0, 38.2, 38.0, 38.9, 38.8, 38.2, 38.5, 38.4, 39.1, 38.7
    mean (MB): 38.6
    median (MB): 38.6
  after:
    build times (ms): 815, 816, 832, 832, 826, 809, 823, 806, 825, 829
    mean (ms): 821
    median (ms): 824
    used memory (MB): 28.2, 32.8, 32.9, 32.5, 33.0, 27.4, 32.6, 32.7, 28.9, 33.0
    mean (MB): 31.4
    median (MB): 32.6
```

**Does this PR introduce a breaking change?**

no

### CODE REVIEW

File: lib/MultiCompiler.js [hunk 1]
```diff
@@ -30,25 +30,27 @@ module.exports = class MultiCompiler extends Tapable {
 		this.compilers = compilers;
 		let doneCompilers = 0;
 		let compilerStats = [];
-		this.compilers.forEach((compiler, idx) => {
+		let index = 0;
+		for(const compiler of this.compilers) {
 			let compilerDone = false;
-			compiler.hooks.done.tap("MultiCompiler", stats => {
+			let compilerIndex = index++;
```
Comment (sokra) [inline lib/MultiCompiler.js]: const

File: lib/optimize/FlagIncludedChunksPlugin.js [hunk 2]
```diff
@@ -9,25 +9,25 @@ class FlagIncludedChunksPlugin {
 	apply(compiler) {
 		compiler.hooks.compilation.tap("FlagIncludedChunksPlugin", (compilation) => {
 			compilation.hooks.optimizeChunkIds.tap("FlagIncludedChunksPlugin", (chunks) => {
-				chunks.forEach((chunkA) => {
-					chunks.forEach((chunkB) => {
+				for(const chunkA of chunks) {
+					loopB: for(const chunkB of chunks) {
```
Comment (sokra) [inline lib/optimize/FlagIncludedChunksPlugin.js]: Label is not needed. A simple `continue` would be enough

**Comment by sokra** [comment]
Elsewise fine.

File: lib/optimize/FlagIncludedChunksPlugin.js (hunk 2 above)
Comment (ooflorent) [reply lib/optimize/FlagIncludedChunksPlugin.js]: It is because of lines 25-27.

Comment (sokra) [reply lib/optimize/FlagIncludedChunksPlugin.js]: ah I see

**Comment by sokra** [comment]
Thanks

**Pull request merged**